### ISSUE TITLE
DNS federation API doesn't support changing resource records on Route53

### ISSUE THREAD

**Issue by justinsb** [kubernetes/kubernetes#28477]
I'm writing a controller to set DNS hostnames though annotations, and I get an error while trying to change an existing record, when using AWS Route53

```
 error updating records for {A api.internal.jul4a.mydoomain.com.}: error creating resource record api.internal.jul4a.mydomain.com. A: InvalidChangeBatch: Tried to create resource record set [name='api.internal.jul4a.mydomain.com.', type='A'] but it already exists
        status code: 400, request id: b3bc9474-419d-11e6-9018-170214ac02d2
```

I believe this is because we are using CREATE as the action, but should probably be using UPSERT.  I don't know if I'm using the API wrong though (I don't _think_ I should be deleting the previous record first, for example).

cc @quinton-hoole

(otherwise the API is great quinton - really easy to use!)

**Comment by quinton-hoole** [comment]
Hi @justinsb 

The dnsprovider interface for resource record sets is still pretty basic, and only supports add and remove operations right now:

https://github.com/kubernetes/kubernetes/blob/master/federation/pkg/dnsprovider/dns.go#L46

We could add an update method quite easily, if it's important?

Q

**Comment by justinsb** [comment]
It's important to me - important enough that I can contribute one :-)

I created a dns controller for k8s, so I need to be able to update the DNS records as members come and go.  I'm using it in kops for registering master nodes, so they are actually relatively static, but I still need to fix it:
https://github.com/kubernetes/kops/pull/26

**Comment by quinton-hoole** [comment]
@justinsb What we're doing elsewhere is simply doing a Remove followed by an Add instead of an Update operation.  Will that not work for you here?  Of course an Update would be a bit better, but only marginally so.

**Comment by justinsb** [comment]
The problem is that there's a window where we have no record, I think?  And at least on AWS an immediate add will fail because the delete will still be propagating, so it's a non-trivial window.

I'm playing around with building a `ResourceRecordChangeset`, and moving the Add and Remove methods to it, and then adding an Apply method on the changeset.  The idea is that on AWS & GCE it will be atomic and immediate.  On other providers it might not be guaranteed atomic, but if the provider supports update, then the provider can change a Delete & Add to an update.

Also, changes to the callers are minimal ... they can continue to do atomic removes followed by atomic adds if they want to :-)

The problem I hit with the "Replace" method is that GCE requires you to specify the "before" version as a Delete.  So the caller has to express it as a Delete & an Add for GCE.

**Comment by quinton-hoole** [comment]
@justinsb I think the right way to do it is:

1. Provide an Update() method, which takes the old and the new state as inputs.  This alone will provide the atomic update functionality you're looking for, and should work consistently across all DNS providers.
2. Separately, provide support for multi-operation transactions.  This we could do in one of two ways:
    1. Provide BeginTransaction() and CommitTransaction() methods (with all operations between the two being subject to surrounding transaction, or perhaps supporting an optional TransactionID to associate them with the transaction.
    2. Provide a CompoundOperation() method, which takes as input an array of Addition, Removal and Update operations, which all execute together, atomically.

I suggest that we do the Update() part in 1 above first to unblock you, and then tackle 2 (Transactions) as a separate exercise.  Sound good?

**Comment by justinsb** [comment]
Totally untested, but I just put up #28668 to show how the API could work.  At least for GCE & AWS DNS, both expose the changeset idea, so it feels pretty natural to just go straight to using that.  But I'm not sure whether e.g. you have another DNS provider in mind...

I think the changeset API is simple enough that we could just switch the API over (i.e. no reason to create more methods)

**Comment by quinton-hoole** [comment]
Addressed by  #28668 . Closing.
